File: examples/cpp/benchmark_large_2d.cpp
/**
 * Benchmark large 2-D MOLE operators directly in C++.
 */

#include "mole.h"

#include <algorithm>
#include <chrono>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>


struct BenchmarkResult {
    double median_seconds;
    arma::uword rows;
    arma::uword columns;
    arma::uword nonzeros;
};


double median(std::vector<double> values)
{
    std::sort(values.begin(), values.end());

    const std::size_t size = values.size();
    const std::size_t middle = size / 2;

    if (size % 2 == 0) {
        return 0.5 * (values[middle - 1] + values[middle]);
    }

    return values[middle];
}


template <typename Constructor>
BenchmarkResult benchmark(
    const std::string& name,
    Constructor constructor,
    int repeats
)
{
    std::vector<double> times;

    arma::uword rows = 0;
    arma::uword columns = 0;
    arma::uword nonzeros = 0;

    for (int repetition = 0; repetition < repeats; ++repetition) {
        const auto start = std::chrono::steady_clock::now();

        auto matrix = constructor();

        const auto finish = std::chrono::steady_clock::now();

        const std::chrono::duration<double> elapsed = finish - start;
        times.push_back(elapsed.count());

        rows = matrix.n_rows;
        columns = matrix.n_cols;
        nonzeros = matrix.n_nonzero;

        // Read matrix information so the compiler cannot discard construction.
        volatile arma::uword observed_nonzeros = matrix.n_nonzero;
        (void) observed_nonzeros;
    }

    const double median_seconds = median(times);

    std::cout
        << std::left << std::setw(28) << name
        << " shape=(" << rows << ", " << columns << ")"
        << " nnz=" << std::setw(12) << nonzeros
        << " median=" << std::fixed << std::setprecision(6)
        << median_seconds << " s"
        << " runs=[";

    for (std::size_t index = 0; index < times.size(); ++index) {
        if (index > 0) {
            std::cout << ", ";
        }

        std::cout << std::fixed << std::setprecision(6) << times[index];
    }

    std::cout << "]\n";

    return {
        median_seconds,
        rows,
        columns,
        nonzeros
    };
}


int main(int argc, char** argv)
{
    const u32 cells =
        argc > 1 ? static_cast<u32>(std::stoul(argv[1])) : 500;

    const u16 order =
        argc > 2 ? static_cast<u16>(std::stoul(argv[2])) : 4;

    const int repeats =
        argc > 3 ? std::stoi(argv[3]) : 3;

    const Real dx = 1.0 / static_cast<Real>(cells);
    const Real dy = 1.0 / static_cast<Real>(cells);

    arma::ivec dirichlet(4, arma::fill::ones);
    arma::ivec neumann(4, arma::fill::zeros);

    std::cout << "\nMOLE C++ large 2-D benchmark\n";
    std::cout << "Order:       " << order << '\n';
    std::cout << "Cells:       " << cells << " x " << cells << '\n';
    std::cout << "Spacing:     " << dx << " x " << dy << '\n';
    std::cout << "Repetitions: " << repeats << "\n\n";

    std::cout
        << "C++ timings include operator construction only; "
        << "there is no conversion to SciPy.\n\n";

    std::vector<std::pair<std::string, double>> results;

    auto laplacian_result = benchmark(
        "laplacian_2d",
        [=]() {
            return Laplacian(
                order,
                cells,
                cells,
                dx,
                dy
            );
        },
        repeats
    );

    results.push_back({
        "laplacian_2d",
        laplacian_result.median_seconds
    });

    auto gradient_result = benchmark(
        "gradient_2d",
        [=]() {
            return Gradient(
                order,
                cells,
                cells,
                dx,
                dy
            );
        },
        repeats
    );

    results.push_back({
        "gradient_2d",
        gradient_result.median_seconds
    });

    auto divergence_result = benchmark(
        "divergence_2d",
        [=]() {
            return Divergence(
                order,
                cells,
                cells,
                dx,
                dy
            );
        },
        repeats
    );

    results.push_back({
        "divergence_2d",
        divergence_result.median_seconds
    });

    auto interpolation_result = benchmark(
        "interpolation_2d",
        [=]() {
            return Interpol(
                cells,
                cells,
                0.5,
                0.5
            );
        },
        repeats
    );

    results.push_back({
        "interpolation_2d",
        interpolation_result.median_seconds
    });

    auto cell_to_face_result = benchmark(
        "interpolate_cell_to_face_2d",
        [=]() {
            return InterpolCtoF(
                order,
                cells,
                cells,
                dirichlet,
                neumann
            );
        },
        repeats
    );

    results.push_back({
        "cell_to_face_2d",
        cell_to_face_result.median_seconds
    });

    std::cout << "\nCSV summary\n";
    std::cout << "operation,cpp_seconds\n";

    for (const auto& result : results) {
        std::cout
            << result.first << ','
            << std::fixed << std::setprecision(9)
            << result.second << '\n';
    }

    return EXIT_SUCCESS;
}

File: examples/python/benchmark_large_2d.py
"""Benchmark large 2-D MOLE operators through the Python interface."""

import argparse
import gc
import statistics
import time

import mole
import numpy as np
from scipy.sparse import csr_matrix


def benchmark(name, constructor, repeats):
    """Measure construction plus C++-to-SciPy conversion time."""

    times = []
    shape = None
    nonzeros = None

    for _ in range(repeats):
        gc.collect()

        start = time.perf_counter()
        matrix = constructor()
        elapsed = time.perf_counter() - start

        assert isinstance(matrix, csr_matrix)

        times.append(elapsed)
        shape = matrix.shape
        nonzeros = matrix.nnz

        # Make sure the matrix is actually usable.
        vector = np.ones(matrix.shape[1])
        result = matrix @ vector
        assert np.all(np.isfinite(result))

        del result
        del vector
        del matrix

    median = statistics.median(times)

    print(
        f"{name:28s}"
        f" shape={str(shape):20s}"
        f" nnz={nonzeros:12,d}"
        f" median={median:10.6f} s"
        f" runs={[round(value, 6) for value in times]}"
    )

    return median


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--cells", type=int, default=500)
    parser.add_argument("--order", type=int, default=4)
    parser.add_argument("--repeats", type=int, default=3)
    arguments = parser.parse_args()

    cells = arguments.cells
    order = arguments.order
    repeats = arguments.repeats

    dx = 1.0 / cells
    dy = 1.0 / cells

    dirichlet = np.ones(4, dtype=np.int64)
    neumann = np.zeros(4, dtype=np.int64)

    print("\nMOLE Python large 2-D benchmark")
    print(f"Order:      {order}")
    print(f"Cells:      {cells} x {cells}")
    print(f"Spacing:    {dx} x {dy}")
    print(f"Repetitions:{repeats}")
    print()
    print(
        "Python timings include both C++ construction and conversion "
        "to SciPy CSR.\n"
    )

    # Small warm-up calls load SciPy and initialize the extension before
    # measuring the large matrices.
    mole.laplacian_1d(2, 10, 0.1)
    mole.interpolation_1d(10, 0.5)

    results = {}

    results["laplacian_2d"] = benchmark(
        "laplacian_2d",
        lambda: mole.laplacian_2d(
            order,
            cells,
            cells,
            dx,
            dy,
        ),
        repeats,
    )

    results["gradient_2d"] = benchmark(
        "gradient_2d",
        lambda: mole.gradient_2d(
            order,
            cells,
            cells,
            dx,
            dy,
        ),
        repeats,
    )

    results["divergence_2d"] = benchmark(
        "divergence_2d",
        lambda: mole.divergence_2d(
            order,
            cells,
            cells,
            dx,
            dy,
        ),
        repeats,
    )

    results["interpolation_2d"] = benchmark(
        "interpolation_2d",
        lambda: mole.interpolation_2d(
            cells,
            cells,
            0.5,
            0.5,
        ),
        repeats,
    )

    results["cell_to_face_2d"] = benchmark(
        "interpolate_cell_to_face_2d",
        lambda: mole.interpolate_cell_to_face_2d(
            order,
            cells,
            cells,
            dirichlet,
            neumann,
        ),
        repeats,
    )

    print("\nCSV summary")
    print("operation,python_seconds")

    for name, elapsed in results.items():
        print(f"{name},{elapsed:.9f}")


if __name__ == "__main__":
    main()